src/protocols/bdx/AsyncTransferFacilitator.h

```diff
@@ -0,0 +1,136 @@
+    TransferSession mTransfer;
+
+private:
+    bool mProcessingOutputEvents = false;
+
+    // The Exchange holder that holds the exchange context used for sending and receiving BDX messages.
+    Messaging::ExchangeHolder mExchange;
+
+    // The timeout for the BDX transfer session.
+    System::Clock::Timeout mTimeout;
+
+    System::Layer * mSystemLayer;
+};
+
+/**
+ * An AsyncTransferFacilitator that is initialized to respond to an incoming BDX transfer request.
+ * An AsyncResponder object is associated with an exchange and handles all BDX messages sent over that exchange.
+ *
+ * Provides a method for initializing the TransferSession members but still needs to be extended to implement
+ * HandleTransferSessionOutput.
+ *
+ * An instance of some subclass of this class should be used as the exchange delegate for a BDX transfer.
+ */
+class AsyncResponder : public AsyncTransferFacilitator
+{
+public:
+    /**
+     * Initialize the TransferSession state machine to be ready for an incoming transfer request.
+     *
+     * @param[in] exchangeCtx     The exchange to use for the transfer.
+     * @param[in] role            The role of the Responder: Sender or Receiver of BDX data
+     * @param[in] xferControlOpts Supported transfer modes (see TransferControlFlags)
+     * @param[in] maxBlockSize    The maximum supported size of BDX Block data
+     * @param[in] timeout         The chosen timeout delay for the BDX transfer
+     */
+    CHIP_ERROR Init(System::Layer * layer, Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                    BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize, System::Clock::Timeout timeout);
+
+    /**
+     * This is called by the subclass implementing HandleTransferSessionOutput to notify the AsyncResponder
+     * that it has handled the OutputEvent specified in "event" and "status" is the result of handling the event.
+     * If the OutputEvent was handled successfully by the subclass that means the TransferSession API call to handle the output
+     * event was successful and the TransferSession has the next message prepared. In that case, the "status" is set to
+     * CHIP_NO_ERROR and the AsyncResponder should call ProcessOutputEvents to get and process the next output events from the
+     * TransferSession to proceed with the BDX transfer. Otherwise if the output event indicates the end of BDX transfer (like EOF
+     * receieved or transfer timeout), it will clean up and destroy itself and the subclass. If the "status" is set to an error
+     * code, the AsyncResponder should abort the transfer and clean up once the status report is sent over the exchange.
+     *
+     * @param[in] event The OutputEvent that was handled by the subclass.
+     * @param[in] status The error code that occured when handling the event if an error occurs. Otherwise CHIP_NO_ERROR.
+     */
+    void NotifyEventHandled(TransferSession::OutputEvent & event, CHIP_ERROR status);
```

Comment:
This should be const, and ToString on OutputEvent should be a static method, since it does not depend on any state of the OutputEvent itself.